lib/PhasarLLVM/TypeHierarchy/DIBasedTypeHierarchy.cpp

```diff
@@ -0,0 +1,323 @@
+        assert(TransitiveClosure.size() >= ActualDerivedType);
+        TransitiveClosure[BaseTypeVertex][ActualDerivedType] = true;
+
+        continue;
+      }
+    }
+  }
+
+  // Add transitive edges
+  bool Change = true;
+  size_t TCSize = TransitiveClosure.size();
+  size_t RowIndex = 0;
+  size_t PreviousRow = 0;
+
+  // (max): I know the code below is very ugly, but I wanted to avoid recursion.
+  // If the algorithm goes over the entire matrix and couldn't update any rows
+  // anymore, it stops. As soon as one position gets updated, it goes over the
+  // matrix again
+  while (Change) {
+    Change = false;
+    // go over all rows of the matrix
+    for (size_t CurrentRow = 0; CurrentRow < TCSize; CurrentRow++) {
+      // compare current row with all other rows and check if an edge can be
+      // added
+      for (size_t CompareRow = 0; CompareRow < TCSize; CompareRow++) {
+        // row doesn't need to compare itself with itself
+        if (CurrentRow == CompareRow) {
+          continue;
+        }
+
+        // if the current row is not a parent type of the current compare row,
+        // no edges should be added
+        if (!TransitiveClosure[CurrentRow][CompareRow]) {
+          continue;
+        }
+
+        // Compare both rows and add edges if needed
+        for (size_t Column = 0; Column < TCSize; Column++) {
+          if (TransitiveClosure[CompareRow][Column] &&
+              !TransitiveClosure[CurrentRow][Column] && CurrentRow != Column) {
+            TransitiveClosure[CurrentRow][Column] = true;
+            Change = true;
+          }
+        }
+      }
+    }
+  }
+
+  // add edges onto vertices themselves
+  for (size_t I = 0; I < TransitiveClosure.size(); I++) {
+    TransitiveClosure[I][I] = true;
+  }
+
+  // add virtual functions table
+  size_t VTableSize = 0;
+  for (const auto &Subprogram : Finder.subprograms()) {
+    if (Subprogram->getVirtualIndex() > VTableSize) {
+      VTableSize = Subprogram->getVirtualIndex();
+    }
+  }
+  // if the biggest virtual index is 2 for example, the vector needs to have a
+  // size of 3 (Indices: 0, 1, 2)
+  VTableSize++;
+
+  std::vector<const llvm::Function *> Init = {};
+  std::vector<std::vector<const llvm::Function *>> IndexToFunctions;
+  for (size_t I = 0; I < VTableSize; I++) {
+    IndexToFunctions.push_back(Init);
+  }
+  // get VTables
+  for (auto *Subprogram : Finder.subprograms()) {
+    if (!Subprogram->getVirtuality()) {
+      continue;
+    }
+
+    const auto *const FunctionToAdd =
+        IRDB.getFunction(Subprogram->getLinkageName());
+
+    if (!FunctionToAdd) {
+      continue;
+    }
+
+    const auto &TypeIndex = TypeToVertex.find(Subprogram->getContainingType());
+
+    if (TypeIndex->getSecond() >= IndexToFunctions.size()) {
+      continue;
+    }
+
+    IndexToFunctions[TypeIndex->getSecond()].push_back(FunctionToAdd);
+  }
+
+  for (auto &ToAdd : IndexToFunctions) {
+    VTables.emplace_back(std::move(ToAdd));
+  }
```

Comment:
This looks wrong. The indices of the vtables must match the associated type index